add node-oracledb support

atiertant opened the pull request:
this look like working,could someone test?
thanks @KhalilTN

rhys-vdw commented:
> this look like working

These files must be created in the `/src` directory and compiled to generate `/lib`.

> could someone test?

I don't know much about oracle or oracledb, so hopefully someone more knowledgable can weight in.

jgoux commented:
Hello,
any chance this getting merged ?

elhigu commented:
There are basically no other tests, but only schema building for this. Maybe you could still add test oracledb client to `test/unit/query/builder.js` and make sure that queries are generated correctly, especially in those cases where oracledb driver works differently from base implementation.

atiertant commented:
@elhigu done ! i'm writting and sql adapter for waterline using this code for oracle,all test passed except blob because oracledb driver has a curious way to insert blob... they plan to simplify it and i'll update my code then.

elhigu commented:
I think this looks fine (from a viewpoint not knowing anything about those dialects). 

Only thing that is bothering me is that it looks mostly copy-paste of the old oracle client. There might be possibility to share codebase of the clients. e.g. websql is sqlite3 derivate so that could be used as an example.

However this is not breaking anything and I suppose it works... so it would be ok for me to merge this as is and do the refactoring later on :) @rhys-vdw  what do you think?

rhys-vdw commented:
Hey @atiertant, thanks for the PR. It's challenging to review this PR because the changes are hidden in the duplicated code. Using inheritance here makes more sense than copy paste - not just for maintainability, but also for readability. The diff should show just the changes that pertain to OracleDB.

Ideally we'd do something like creating an `AbstractOracleClient` (supporting all the Oracle stuff) and then have each Oracle dialect inherit from this. Each inheriting Oracle dialect module should be quite thin and just override the sections pertaining to the API of its oracle module.

I understand the pressing need for this support - but I don't want to merge a PR containing wholesale duplication of files. There has to be clean up here before the merge. Do you think this will be possible, @atiertant?

atiertant commented:
@rhys-vdw hi, there is only one oracle database,and there were many drivers before oracle start writting their own driver for their database.
why would you like to maintain a dialect for an old and no more maintained driver (`This library is not maintained. Oracle has made there own driver.` see https://github.com/joeferner/node-oracle) ? i could move this changes directly in oracle dialect if you want... what do you think of it?

vschoettke commented:
@atiertant strong-oracle is except for the initialization identical to the old oracle driver.

atiertant commented:
so this should be correct ...

rhys-vdw commented:
Hey @atiertant, thanks for your patience.

> why would you like to maintain a dialect for an old and no more maintained driver 
I wasn't aware that the project was unmaintained until now - cheers. It's a good point.

> i could move this changes directly in oracle dialect if you want... what do you think of it?

I would support using `oracle` as the client name for `oracledb` and renaming the unsupported driver to `node-oracle` - that way it seems more 'default'. But why remove the support? We have the module and there's no harm in keeping it (for now at least).

> so this should be correct ...

Take a [look](https://github.com/tgriesser/knex/blob/master/src/dialects/strong-oracle/index.js). It inherits from the Oracle driver.

I can't see any benefit in introducing that much duplication into the database. It's a liability. It's not fair on future contributors to have to know to make fixes in two places - and it's more work for collaborators. (I'm not really sure what you've changed because it's just a huge copy+paste - but on my cursory reading it looked almost identical.)

I appreciate the work, but I can't merge it until it's DRY.

atiertant commented:
@rhys-vdw thanks for your review,

it's true that my first commit duplicated all dialect' code and that was not a good idea...
i removed all unnecessary duplicated code to make futur fixes at only one places.
of course the code staying is a copy paste of the original oracle with some little changes but
keep in mind all current oracle drivers are based on joeferner works and i woudn't like to reinvent the wheel but oracle made a lot of changes and api is now different.

this is the list of changes:

* add bindOutString function to oracle and oracledb formatter dialect (oracledb binding params are differents)
* Client_Oracledb.prototype._driver (driver is not the same)

the way to deal with connections is a bit different:
* Client_Oracledb.prototype.acquireRawConnection
* Client_Oracledb.prototype.destroyRawConnection

the results are returned into a different format:
* Client_Oracledb.prototype._query
* Client_Oracledb.prototype.processResponse

* oracledb driver correct a bug in timestamp so we must use 'timestamp with local time zone' that affect time,datetime and timestamp of the ColumnCompiler.

now all this changes concert only the way the dialect deals with driver.

epoberezkin commented:
:+1:

atiertant commented:
@rhys-vdw does this arguments convince you or should i emulate old driver functions? please let me know the way you want to do it...

jgoux commented:
It seems great ! Good work @atiertant ! :+1: 
Could someone merge it ?

jrop commented:
We are anxious to use this in our project.  Is there any plan to merge this in?

tgriesser commented:
Yep, will look to get this in later this week - thanks for your patience

mrchess commented:
lmk if you think you need more testing on this one. Count me in!

mrchess commented:
@atiertant Are you using this in your branch right now? Want me to use it too to help test? I wanted to use it so I could leverage knex transactions and Oracle :)

atiertant commented:
@mrchess yes we are using this branch, simple request are working, we didn't tryed transactions, i don't know if it's working... we are working on blob insertion.
it would be great to list every test you've done with the status (working or error encountered)

squarejaw commented on lib/dialects/oracledb/index.js:
```diff
@@ -0,0 +1,167 @@
+var ReturningHelper = require('../oracle/utils').ReturningHelper;
+var Promise = require('../../promise');
+var stream = require('stream');
+var async = require('async');
```
`async` is currently in `devDependencies` in the package.json. Since it's being used in the dialect code it should be moved to `dependencies` right?

atiertant commented:
@squarejaw thanks for the review ! i use it only one time... i will remove it.

elhigu commented:
@atiertant could you also fix the conflicts? that way it will be easier to just merge this after the final review :)

serkanberksoy commented:
Guys I really need this fast for a project. Is there a workaround to get it? Can I build it from a branch?

vschoettke commented:
@serkanberksoy the strong-oracle dialect might still be an option for you.

atiertant commented:
@serkanberksoy if you want use oracle corporate driver, you could use https://github.com/Atlantis-Software/knex.git for now blob can't be insert or update, we are working on it...

atiertant commented:
now we can insert blob !
@squarejaw: async removed. 
we are now working on update with blob.

jrop commented:
@atiertant We are having an issue where we get this error: `Cannot read property 'rowsAffected' of undefined` when inserting rows into a table. On line 139 of `lib/dialects/oracledb/index.js`, if we change it using this patch, then the insert works for us:

```diff
--- a/lib/dialects/oracledb/index.js
+++ b/lib/dialects/oracledb/index.js
@@ -136,7 +136,7 @@
     return connection.executeAsync(obj.returningSql, rowIds);
   }).then(function (response) {
     obj.response = response.rows;
-    obj.rowsAffected = response.rows.rowsAffected;
+    obj.rowsAffected = response.rowsAffected;
     return obj;
   });
 };
```

atiertant commented:
@jrop thanks for your tests but this bug as already been [corrected](https://github.com/tgriesser/knex/pull/990/files#diff-d7aec0cda7cf0e6122d01d65253c3f40R137)

please

    git clone https://github.com/Atlantis-Software/knex.git
    cd knex
    npm install
    npm run babel

before use to recompil

lumberj commented:
Hey, anyway I can help?

atiertant commented:
@lumberj yes you can test this branch and give us feedback.
@elhigu confilct solved, a travis job should be restart to pass (it's a mysql error)

now update and insert works with blob and returning !
next job transaction...

lumberj commented:
@atiertant this should work right?

```
npm install Atlantis-Software/knex
```

lumberj commented:
Pulled down this branch and ran `npm install` then `npm link` which should run `build`. I get this error at the end:

```
WARNING in ./~/oracledb/lib/oracledb.js
Module not found: Error: Cannot resolve 'file' or 'directory' ../build/Release/oracledb in /Users/alan/Projects/knex/node_modules/oracledb/lib
 @ ./~/oracledb/lib/oracledb.js 27:18-54

ERROR in ./~/oracledb/lib/oracledb.js
Module not found: Error: Cannot resolve 'file' or 'directory' ../build/Debug/oracledb in /Users/alan/Projects/knex/node_modules/oracledb/lib
 @ ./~/oracledb/lib/oracledb.js 30:19-53
```

Also tried running unit tests for github.com/interledger/five-bells-ledger with this branch. I didn't expect it to work, and it didn't:

```
Knex:warning - Can't take lock to run migrations: connection.setAutoCommit is not a function
Knex:warning - If you are sure migrations are not running you can release the lock manually by deleting all the rows from migrations lock table: migrations_lock
```

The remaining db tests fail.

CorvusCorrax commented:
Hi lumberj !
Your first error sounds like the build didn't work, try : 
`npm run babel`
As for your second problem, it seems that you use transactions, which are not yet supported, but i'm working on it, it should be available soon.

atiertant commented:
@lumberj `npm install Atlantis-Software/knex` doesn't work.
you should use: 
```
git clone https://github.com/Atlantis-Software/knex.git
cd knex
npm install
npm run babel
```

if you wouldn't like install git you could download [master.zip](https://github.com/Atlantis-Software/knex/archive/master.zip) then extract and : 
```
cd knex
npm install
npm run babel
```

rhys-vdw commented:
Hey @atiertant, sorry I've been unresponsive. I'm assuming the PR is still ready to be merged?

atiertant commented:
@rhys-vdw ,it's ok. yes this could be merged, now we only commit src to avoid conflicts but i suppose `npm run babel` is done before publish...
we are now working on transactions and optimising returning sql but this could be done in an other PR.

atiertant commented:
now all is working with transaction and lob.
@rhys-vdw this is now fully ready to merge, all file has been commit without any conflict.

lumberj commented:
Just tried running again with the latest changes. This error message starts streaming in the terminal:

```
Knex:warning - Pool2 - Error: ORA-00923: FROM keyword not found where expected
```

Running against:
https://github.com/interledger/five-bells-ledger

atiertant commented:
@lumberj you re right ! there was a bug on oracle dialect in tgriesser/knex master i rebased, in https://github.com/tgriesser/knex/pull/1296 that as been merged.
the pool ping function do `SELECT 1` that should be change to `SELECT 1 FROM DUAL`, we fixed it at https://github.com/Atlantis-Software/knex/pull/7/commits/0da1c773a01db85fa9a609ef9265e6d564fc5896#diff-c8ff07bea2af41f7c0247b860dbd9d39L174
@elhigu could you restart the job in travis (one more time mysql error)

lumberj commented:
Tests passing for me now

AiNoKame commented:
That's so much for this @atiertant ! Quick question: I need to use [fetchAsString](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#propdbfetchasstring) for oracledb.DATE and oracledb.NUMBER, but is there a way to configure this? Right now, I'm adding:

`oracledb.fetchAsString = [ oracledb.DATE, oracledb.NUMBER ];` to https://github.com/tgriesser/knex/pull/990/files#diff-a86ace7e0baa34caa6f981181dc544e2R32 and it works great

atiertant commented:
@AiNoKame knex return Date object for date and number for number ...maybe using fetchAsString is not the only way to do what you want. couldn't you use `.toString()` to data before use?
this code doesn't pass any options from config to driver, i don't know if other dialect do...
this is out of this PR scope, but you could propose a PR when this one has been merged.

AiNoKame commented:
@atiertant Sure thing! I just want to list some use cases for wanting to have the option to configure the use of fetchAsString for Date and Number as core functionality:

Date - Currently, node-oracledb only supports two Oracle Date datatypes: `DATE` and `TIMESTAMP`, both of which don't have a timezone (:scream:). If your oracle schema has columns of types `TIMESTAMP WITH TIMEZONE` or `TIMESTAMP WITH LOCAL TIMEZONE`, a node-oracldedb fetch with that column selected just fails with an unsupported datatype error (`Error: select * from "db_schema_versions" order by "created_at" desc - NJS-010: unsupported data type in select list`), so the only current workaround is using fetchAsString.

Number - Here's an issue that @lumberj posted at strong-oracle's github: https://github.com/strongloop/strong-oracle/issues/63. Databases that support values with 16+ significant digits, when fetched from the database as a Number causes a Javascript rounding error.

As for passing options from config to driver, I'm using `connection.requestTimeout` for mssql, so it can work :smile_cat: Although, I tried doing a `const oracleDb = require('oracledb')` and adding `connection.fetchAsString = [ oracledb.DATE, oracledb.NUMBER ]` to my oracle connection config, but it didn't work :cry: 

(Sorry for the long post!)

vschoettke commented:
@AiNoKame There is still the option to use the oracle cast function e.g. `CAST(<COLNAME> as varchar2(<count>))` in the query to convert the unsupported/overflowing column to a string.

atiertant commented:
@AiNoKame it looks like all date are fetched as `TIMESTAMP WITH LOCAL TIMEZONE`, look at doc [result-type-mapping](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#-916-result-type-mapping).
for `const oracleDb = require('oracledb')`, i will look at mssql code...

codyzu commented:
Is this still in progress? I'm very interested :-)

atiertant commented:
@codyzu we are now testing this in our projects

codyzu commented:
@atiertant great! glad to hear it.

The failure on travis seems to be a timeout error (related to mysql). Do you think this will be merged into the master (and installable with npm) any time soon???

Do you need help?

atiertant commented:
@codyzu we need some time to test transactions but all other features are working.

rhys-vdw commented:
@atiertant When you're ready, please rebase your changes and remove `/lib` etc? We no longer track the built files.

atiertant commented:
@rhys-vdw do you mean `git rm /lib` ?

codyzu commented:
@rhys-vdw @atiertant Sorry to be a pain, but I really want to use this in our project. Any news? Do you need some help?

As mentioned in the comments above, we can't install directly with npm due to the babel compile... I'm thinking of cloning the entire branch, and either adding a postinstall script to package.json to do the babel compile or checking in the compiled sources. This would allow my project to use the branch as is (by pointing npm to my repo). Good idea? Of course, I would rather see this PR merged... so don't hesitate to ask if you need help!

jrop commented:
I believe the way to do this is to only check in the uncompiled sources to
git, and then add a prepublish hook to package.json that compiles the
sources and only publishes the result to npm.  But build artifacts should
generally not be checked in to source control.
On May 27, 2016 4:12 AM, "Cody Zuschlag" <notifications@github.com> wrote:

> @rhys-vdw <https://github.com/rhys-vdw> @atiertant
> <https://github.com/atiertant> Sorry to be a pain, but I really want to
> use this in our project. Any news? Do you need some help?
>
> As mentioned in the comments above, we can't install directly with npm due
> to the babel compile... I'm thinking of cloning the entire branch, and
> either adding a postinstall script to package.json to do the babel compile
> or checking in the compiled sources. This would allow my project to use the
> branch as is (by pointing npm to my repo). Good idea? Of course, I would
> rather see this PR merged... so don't hesitate to ask if you need help!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgriesser/knex/pull/990#issuecomment-222098937>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ABVslY2lf6PkpN_1SR5mZq4kU1JGwNxyks5qFrWVgaJpZM4GBijT>
> .
>

atiertant commented:
@codyzu we are working on transactions and internal tests, we hope finnish it next week.
[the problem in installation with node-oracledb](https://github.com/oracle/node-oracledb/issues/424) could be solved by installing node-oracledb in global mode : `npm install -g oracledb` or install it after knex. you could clone our branch to start your dev but i think it it would be better to wait merge in knex for production environment.

williamgueiros commented:
@atiertant first of all , tanks for you hard work on this feature.

do you need help to finish the PR?

atiertant commented:
@rhys-vdw ready to merge !

rhys-vdw commented:
Awesome. A lot of people excited about this. I'm still a bit intimidated by the support issue with the duplication, but I'm happy to put that aside in the interest of getting this popular feature merged.

So I've just checked out this branch and it's hitting over 100 linter errors. **We need 0 errors, 0 warnings before merging.** I used [esnext](https://github.com/esnext/esnext/blob/master/src/esnext.js), but it's mostly replacement of `var` with `let`/`const` as appropriate so it might be fine to do it by hand.

Then this branch should be rebased onto master. The commit history is also full of noise (merge commits etc, bad commit messages like 'fix bugs'), so it should be squashed into a single (or at least fewer) commits.

I have no access to Oracle, but thanks to @williamgueiros in #1491, we should have the ability to test on Travis. So we can pull in his PR and see those tests run live.

So, steps in summary (anybody is able to do this btw, not just @atiertant):

 1. Create a new PR with a single (or at least fewer) commits rebased onto current master HEAD.
 2. Correct for linter warnings.
 3. Merge in williamgueiros/knex#1
 4. Ensure tests run successfully
 5. Merge

rhys-vdw commented:
PS. If somebody is planning to tackle the above task - please make sure to comment here first to avoid double handling.

ramirezd42 commented:
I'd be happy to clean up the linter errors if someone else isn't already
doing so.

On Tuesday, June 14, 2016, Rhys van der Waerden <notifications@github.com>
wrote:

> PS. If somebody is planning to tackle the above task - please make sure to
> comment here first to avoid double handling.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/tgriesser/knex/pull/990#issuecomment-225843640>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ACGSZkW2mLmvOhPqkY7q2am8AWhTQz3Mks5qLoRMgaJpZM4GBijT>
> .
>

rhys-vdw commented:
@ramirezd42 Great. I think we can safely assume that nobody is doing it at this stage. Perhaps just fork from [atiertant's branch](https://github.com/Atlantis-Software/knex) and merge in master from knex. We can sort out squashing etc later.

ramirezd42 commented:
I fixed all of the linter issues in a PR to @atiertant 's fork here: https://github.com/Atlantis-Software/knex/pull/11. 

I can make a separate PR if necessary, but it seemed easier to just wait for @atiertant to accept changes into that PR to keep the conversation in this thread.

All of the tests pass except for one mariadb test which didn't pass for me on tgriesser/knex master either so I can't imagine it was related:

`1) Integration Tests mariadb | mariasql "before all" hook:
     Error: Pool is destroyed
      at Pool.acquire (node_modules/pool2/lib/pool.js:166:12)
      at ../src/client.js:2:32
      at Client.acquireConnection (../src/client.js:2:32)
      at ../src/runner.js:1:41
  From previous event:
      at ../src/runner.js:1:41
  From previous event:
      at Runner.ensureConnection (../src/runner.js:1:41)
      at Runner.run (../src/runner.js:1:41)
      at SchemaBuilder.Target.then (../src/interface.js:2:20)
      at Migrator.forceFreeMigrationsLock (lib/migrate/index.js:9:6636)
      at Context.<anonymous> (test/integration/migrate/index.js:15:25)`

rhys-vdw commented:
> I fixed all of the linter issues in a PR to @atiertant 's fork

Excellent. Nice one.

> All of the tests pass except for one mariadb test which didn't pass for me on tgriesser/knex master either so I can't imagine it was related

All good. I think that test will pass if we keep re-running the tests. (That's a problem, but not related to this PR).

We're at a point now where I could just pull all this stuff in manually and merge in the CI test branch. But I'll give @atiertant a chance to respond to your PR first.

atiertant commented:
@rhys-vdw so what's next? whould you like to merge in the CI test branch?

rhys-vdw commented:
Hey @atiertant. I want to get #1503 sorted so we know everything's functioning as expected. Hopefully we'll have an elegant solution for testing Oracle soon.

elhigu commented:
Merged this branch as were discussed on contributor chat... If there are issues to be fixed in tests when oracle CI is setup those will be fixed as part of CI enabling pull request. Thanks @atiertant and everyone else 👍

hazembm commented:
Hi for all
i ask if there is an exemple connecting to knex with oracledb.
doesn t work for me
thanks

mrodrig commented:
Any idea when the next release that will include this will be? Would like to use knex for Oracle, but can't until this is released.

elhigu commented:
@mrodrig Looks like we can still do minor release 0.11.8 to release this. I skimmed through latest merges and there didn't seem to be incompatible changes, except bug fixes. Just need to write changelog, bump the version and publish. If I don't have anything important this night I can do it... not making promises though 👍

elhigu commented:
@mrodrig 0.11.8 is now released with this included

mrodrig commented:
Awesome- Thank you @elhigu!
